python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java

```diff
@@ -0,0 +1,264 @@
+    if (typeMutatingMethods == null) {
+      return Collections.emptyList();
+    }
+    return paramSymbol.get().usages().stream().filter(usage -> {
+      Tree parent = usage.tree().parent();
+      if (parent.is(QUALIFIED_EXPR)) {
+        QualifiedExpression qualifiedExpression = (QualifiedExpression) parent;
+        return  getSymbolFromTree(qualifiedExpression.qualifier()).equals(paramSymbol) && isMutatingMethod(typeMutatingMethods, qualifiedExpression.name().name());
+      }
+      return isUsedInDelStatement(usage.tree()) ||
+        isUsedInLhsOfAssignment(usage.tree(), exp -> isAccessingExpression(exp, usage.tree())) ||
+        isUsedInLhsOfCompoundAssignment(usage.tree()) ||
+        isGetItemOnDefaultDict(defaultValueType, usage.tree()) ||
+        mightBeReferencedOutsideOfFunction(usage.tree());
+    }).map(usage -> usage.tree().parent()).collect(Collectors.toList());
+  }
+
+  /**
+   * Detects if shape of tree is equal to
+   * - expression[SOMETHING]
+   * - expression.SOMETHING
+   */
+  private static boolean isAccessingExpression(Expression expression, Tree tree) {
+    return isObjectOfSubscription(tree, expression) || isQualifier(tree, expression);
+  }
+
+  /**
+   *
+   * Detects case where tree might be referenced outside of function
+   * - self.attr = tree
+   */
+  private static boolean mightBeReferencedOutsideOfFunction(Tree tree) {
+    AssignmentStatement assignment = ((AssignmentStatement) TreeUtils.firstAncestorOfKind(tree, ASSIGNMENT_STMT));
+    if (assignment == null) {
+      return false;
+    }
+    return assignment.assignedValue() == tree
+      && assignment.lhsExpressions().stream()
+      .flatMap(expressionList -> expressionList.expressions().stream())
+      .anyMatch(ModifiedParameterValueCheck::isAccessingSelf);
+  }
+
+  private static boolean isAccessingSelf(Expression expression) {
+    switch (expression.getKind()) {
+      case QUALIFIED_EXPR:
+        return isSelf(((QualifiedExpression) expression).qualifier());
+      case SUBSCRIPTION:
+        return isSelf(((SubscriptionExpression) expression).object());
+      default:
+        return false;
+    }
+  }
+
+  private static boolean isSelf(Expression expression) {
+    return expression.is(NAME) && ((Name) expression).name().equals("self");
+  }
+
+  private static List<Tree> getAttributeSet(Symbol paramSymbol) {
+    return paramSymbol.usages().stream()
+      .map(Usage::tree)
+      .filter(tree -> isUsedInLhsOfAssignment(tree, expression -> isQualifier(tree, expression)))
+      .collect(Collectors.toList());
+  }
+
+  private static boolean isUsedInLhsOfCompoundAssignment(Tree tree) {
+    CompoundAssignmentStatement compoundAssignmentStatement = ((CompoundAssignmentStatement) TreeUtils.firstAncestorOfKind(tree, COMPOUND_ASSIGNMENT));
+    if (compoundAssignmentStatement == null) {
+      return false;
+    }
+    return isAccessingExpression(compoundAssignmentStatement.lhsExpression(), tree);
```

Comment:
Nitpick: I would simplify this to `return compoundAssignmentStatement != null && isAccessingExpression(...)`